src/west/commands/project.py

```diff
@@ -868,7 +868,7 @@ def _fetch(project):
     else:
         # The revision is definitely not a SHA, so it's safe to fetch directly.
         # This avoids fetching unnecessary ref space from the remote.
-        fetch_cmd += '{revision}'
+        fetch_cmd += '{revision}:refs/west/{revision}'
```

Comment:
You know, it's funny, I've been wondering why west/xyz refs were being tracked in the above branch but didn't chase it down.

This is technically illegal behavior. We currently promise that we only ever set the `manifest_rev` ref, so the `west/` namespace is not available to us.

We could of course relax that restriction with good reason, but I want to think about what problems this solves a bit.

Reply:
this was done here :https://github.com/zephyrproject-rtos/west/pull/188
in order to solve the issue of a SHA being specified, but that SHA is not available on the remote HEAD (usually master).
If the SHA is located on a branch, we wouldn't have it locally.

Secondly, github does not allow SHAs to be fetch directly.
https://github.com/zephyrproject-rtos/west/issues/139

Reply:
so guess we need to relax this restriction a bit, if we want `west` to work ;)

Reply:
> this was done here :#188

I have a goldfish's memory :). Thanks for the pointer.

> Secondly, github does not allow SHAs to be fetch directly. #139

Right, but I thought this was for refs like pull/x/head, not SHAs, right?

> so guess we need to relax this restriction a bit, if we want west to work ;)

I'm confused about how this helps pull/x/head, since `west/pull/x/head` is a different local ref name for the same commit.

Reply:
maybe I started the confusing by not quoting what I answered.

> You know, it's funny, I've been wondering why west/xyz refs were being tracked in the above branch but didn't chase it down.

The reason for tracking `west/xyz` was PR#188, especially a git does not allow SHAs to be fetched directly, which is described here #139, and therefore we just fetch as PR#140:
```
  if _is_sha(project.revision):
        # Don't fetch a SHA directly, as server may restrict from doing so.
        _git(project, fetch_cmd + ' --tags -- (url)')
        _git(project, 'update-ref (qual-manifest-rev-branch) (revision)')
```
but this only fetch remote HEAD, which is pointing to `master` in this case.
So if the SHA was on a branch, it wouldn't be fetched, which lead to: PR#188.
Where the remote `refs/heads/*` is kept locally under `refs/west/*` and therefore we can checkout any SHA.
You even suggested, just using `refs/west/` https://github.com/zephyrproject-rtos/west/pull/188#discussion_r254361240
With PR#188, we can checkout any SHA.

> This is technically illegal behavior. We currently promise that we only ever set the manifest_rev ref, 

and `manifest-rev` is still the only head we use under the repos normal refs space: `refs/heads/manifest-rev`, nothing has changed here, and I don't propose to change that.
Only proposal is to keep a local refs space under `refs/west/*`  for anything we fetch, so that we can use that later.

> so the west/ namespace is not available to us.

Technically, you're right. `west` is not using the knowledge that a commit is available under `refs/west/*` directly. `west` just checks out the SHA and points `manifetst-rev` to point to that SHA. It doesn't care whether the SHA is available locally, or remotely.
Actually, you can specify any SHA you have locally in your repo, even one that doesn't exists on the remote side, and `west` will still work, regardsless of whether that SHA is available remote or not.

So if we start using this knowledge in https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1437 that is correctly a change in how we use this refs-space.
But as you are only using that refs space `git cat-file` we have not changed the behavior of `west` and we are still not checking out from `refs/west/*`, you are only using it to get information out.

So we are keeping a little more of the remote refs space, locally.
And we are not changing how `west` uses this information, only how an extension command is using that info ;)

Reply:
Right, I understand that git doesn't allow fetching SHAs directly.

However, the point of the `manifest-rev` ref is that it holds a pointer to the latest `revision` in west.yml, even if it is a SHA. So at least for the most recent `west update`, we do not need `west/*` to support SHA revisions.

Now, if the revisions are all SHAs and they dance around as a repository is rebased, we have a problem. Is that what this is solving?

Reply:
> Right, I understand that git doesn't allow fetching SHAs directly.
> 
> However, the point of the `manifest-rev` ref is that it holds a pointer to the latest `revision` in west.yml, even if it is a SHA. So at least for the most recent `west update`, we do not need `west/*` to support SHA revisions.
> 
> Now, if the revisions are all SHAs and they dance around as a repository is rebased, we have a problem. Is that what this is solving?

No, it is solving the issue that a SHA may not be fetched if it is on a remote branch.
Imagine that `west.yml` points to a specific SHA, let's use: 
https://github.com/NordicPlayground/fw-nrfconnect-nrf/commit/618add172a20deb96c45a1135bfd93c569d1955c
in this example, and do what `west` would do without PR#188.

```
$ mkdir /tmp/nrf
$ cd /tmp/nrf
$ git init
$ git fetch --tags -- https://github.com/NordicPlayground/fw-nrfconnect-nrf.git 
remote: Enumerating objects: 28, done.
remote: Counting objects: 100% (28/28), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 19531 (delta 5), reused 8 (delta 3), pack-reused 19503
Receiving objects: 100% (19531/19531), 8.07 MiB | 5.59 MiB/s, done.
Resolving deltas: 100% (13301/13301), done.
From https://github.com/NordicPlayground/fw-nrfconnect-nrf
 * branch              HEAD       -> FETCH_HEAD
...
$ git checkout 9e305316b0d8ef95b0b7c77457a811c75cc48a7e
fatal: reference is not a tree: 9e305316b0d8ef95b0b7c77457a811c75cc48a7e
```
cause the SHA is not on remote HEAD, therefore we fetch the whole refs space into the local ref space `refs/west/*` so that we have all SHAs locally.
As PR#188 adds: `refs/heads/*:refs/west/*`
```
$ git fetch --tags -- https://github.com/NordicPlayground/fw-nrfconnect-nrf.git refs/heads/*:refs/west/*
remote: Enumerating objects: 744, done.
remote: Counting objects: 100% (744/744), done.
remote: Compressing objects: 100% (27/27), done.
remote: Total 1226 (delta 722), reused 723 (delta 716), pack-reused 482
Receiving objects: 100% (1226/1226), 355.50 KiB | 1.09 MiB/s, done.
Resolving deltas: 100% (862/862), completed with 274 local objects.
From https://github.com/NordicPlayground/fw-nrfconnect-nrf
...
$ git checkout 9e305316b0d8ef95b0b7c77457a811c75cc48a7e
Note: checking out '9e305316b0d8ef95b0b7c77457a811c75cc48a7e'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 9e305316 drivers: lte_lc: Fix invalid eDRX default value for NB-IoT
```

I hope this clarifies the issue with SHAs located on branches.

Reply:
> If the SHA is located on a branch [...] So if the SHA was on a branch, [...] No, it is solving the issue that a SHA may not be fetched if it is on a remote branch. [...] I hope this clarifies the issue with SHAs located on branches.

If you had written NON-DEFAULT branch just once I would have understood you 2 or 3 times faster :-)

Reply:
> > Now, if the revisions are all SHAs and they dance around as a repository is rebased, we have a problem. Is that what this is solving?
> 
> No, it is solving the issue that a SHA may not be fetched if it is on a remote branch.
> Imagine that `west.yml` points to a specific SHA, let's use:
> [NordicPlayground/fw-nrfconnect-nrf@618add1](https://github.com/NordicPlayground/fw-nrfconnect-nrf/commit/618add172a20deb96c45a1135bfd93c569d1955c)
> in this example, and do what `west` would do without PR#188.

I'm sorry I was unclear. By "is that what this is solving", I meant "is that what this [pull request #331] is solving"?

I understand the purpose of #188 :).

Reply:
Nice, finally we get back on track discussing the actual content of this PR :)

> I'm sorry I was unclear. By "is that what this is solving", I meant "is that what this [pull request #331] is solving"?

no, the PR #331 is not solving that.
The PR's purpose is stated here: https://github.com/zephyrproject-rtos/west/pull/331#issue-337899162

> To allow later usage of the fetched refs space then git fetch will
always keep a local ref space even fetching non-SHA refs spaces, such
as branches and PRs.

and as we are already keeping remote stuff under `refs/west/*` we might as well keep whatever `west` fetches in there, because then we have it available for later use, and thus may reduce number of additional `git fetches` and hence improved performance.

Note, this is NOT about mirroring the remote refs space, simply just storing the information we anyway have fetched, so that we might avoid fetching the same again.

We are fetching the exact same amount of data with or without PR #331, the PR just have the potential to reduce later redundant fetches.

Reply:
OK, gotcha. Can we make a deal? Will you send a PR to the west documentation saying that west "owns" `manifest-rev` *and* `refs/west/*` before we merge this?

Reply:
>  Will you send a PR to the west documentation saying that west "owns" manifest-rev and refs/west/* before we merge this?

Or at least promise to do so :)

Reply:
Actually, hold on. Sorry, I admit I didn't understand refspecs much until today when I RTFM a bit.

Given `revision: foo` in a manifest file, west expects that `foo` is a SHA, a branch, or a tag.

- if `foo` is a SHA, we are in the previous if block, this line is not executed
- if `foo` is a branch: assuming the remote `.git/foo` and `.git/refs/foo` does not exist, fetching plain `foo` gives us `refs/heads/foo` as desired, and it seems OK to stash that in our local `.git/refs/west/foo`, since we're doing that for SHAs with the `refs/heads/*:refs/heads/*` refspec above
- **but** if `foo` is a tag, say `v1.14.1`, this would create `refs/west/v1.14.1`, which is actually a tag object, created outside of `.git/refs/tags`

Try it:

```
$ git fetch https://github.com/zephyrproject-rtos/zephyr v1.14.0:refs/west/v1.14.0
From https://github.com/zephyrproject-rtos/zephyr
 * [new tag]               v1.14.0    -> refs/west/v1.14.0
$ ls .git/refs/west/v1.14.0
.git/refs/west/v1.14.0
$ git cat-file -t refs/west/v1.14.0
tag
$ git cat-file -t west/v1.14.0
tag
```

But `git tag | grep west` is empty, no results.

Is that really the behavior we want? Serious question, I am really not sure.

Reply:
(And yes, I know we are also fetching with `--tags` -- that's a separate matter.)

Reply:
> since we're doing that for SHAs with the refs/heads/*:refs/heads/* refspec above

I assume you mean `refs/heads/*:refs/west/*` here

>OK, gotcha. Can we make a deal? Will you send a PR to the west documentation saying that west "owns" manifest-rev and refs/west/* before we merge this?

West already own `refs/west/*`, so this would be documenting existing behavior. It's unfortunate that Git doesn't allow us to fetch all refs without actually storing them anywhere, just getting the blobs and objects.

Reply:
> but if foo is a tag,...

I think you meant annotated tag.

> say v1.14.1, this would create refs/west/v1.14.1, which is actually a tag object, created outside of .git/refs/tags

What are the concern(s)? I assumed git would never try to interpret anything outside heads/ and tags/, is this too optimistic? If it too optimistic then maybe just treat tags as a special case and fetch them separately with something equivalent to:  `{revision}^0:refs/west/{revision}`

Or maybe just apply the `^0` treatment to everything fetched, can't hurt...

> It's unfortunate that Git doesn't allow us to fetch all refs without actually storing them anywhere, just getting the blobs and objects.

But then they would go away on the next garbage collection, no?

Reply:
>     * **but** if `foo` is a tag, say `v1.14.1`, this would create `refs/west/v1.14.1`, which is actually a tag object, created outside of `.git/refs/tags`

Partly right, but not fully.
Remember this line in our fetch: https://github.com/zephyrproject-rtos/west/blob/0d1dfa523fc4b06e421f2fc97d4d7c4febb8a063/src/west/commands/project.py#L862
So the `git fetch --tags -- {url} {revision}:refs/west/*` will ensure the tags still goes to `.git/refs/tags` ;)

Reply:
Try for yourself:
```
$ ls .git/refs/tags/                                                                                
$ ls .git/refs/west/                                                                                   
ls: cannot access '.git/refs/west/': No such file or directory                                         
$ git tag                                                                                              
$ git fetch ncs -- v1.1.0:refs/west/v1.1.0                                                             
From https://github.com/NordicPlayground/fw-nrfconnect-nrf                                             
 * [new tag]           v1.1.0     -> refs/west/v1.1.0                                                  
 * [new tag]           v0.1.0     -> v0.1.0                                                            
 .....                                                                                                 
$ git tag                                                          
v0.1.0                                                                                                 
v1.1.0                                                                                                 
....                                                                                                   
$ ls .git/refs/tags/                                                                                   
v0.1.0  ... v1.1.0
$ ls .git/refs/west/
v1.1.0
```
`....` a lot of other tags, removed to reduce noise here.

Reply:
and following up:
```
$ git cat-file -t refs/tags/v1.1.0
tag
```
so is it a problem we keep an additional ref in the `west` ref-space ?

Reply:
> >    Will you send a PR to the west documentation saying that west "owns" manifest-rev and refs/west/* before we merge this?

> Or at least promise to do so :)

I can promise to send a PR describing that `west` also own the `refs/west` space.
But as that PR would be documenting existing behavior, I don't see why it should be coupled to this PR, but that's just my opinion.

Reply:
> What are the concern(s)? 

>  I guess, user will never need to worry or care about our additional `refs/west` space.

We're creating an unbounded amount of un-garbage-collectable refs here, including putting tags outside of refs/tags.

Beyond not being able to collect the refs when they are no longer needed, it just feels weird. I freely admit I'm just sketched out by not understanding the potential implications, and at least personally I'm bothered by all the `west/xyz` output I'm seeing showing up in `git log --oneline` output, for example. (E.g. all the backport branches if you have a downstream where zephyr isn't the manifest repository.)

Within the Nordic downstream, I have always created a tag at the old point in history before rewriting history in any repository managed by west. This ensures an explicit pointer to the ref is available in a place where a user can expect to find it. This feels like going behind the user's back in a weird way.

Reply:
> > What are the concern(s)?
> 
> > I guess, user will never need to worry or care about our additional `refs/west` space.
> 
> We're creating an unbounded amount of un-garbage-collectable refs here, including putting tags outside of refs/tags.
> 

Agree. And a very valid point indeed.
We should look into cleaning up this ref-space.
One option could be to simple wipe the `refs/west/` ref-space upon each `west update`.
That way we ensure remotely deleted branches are removed.
Removed tags are also removed, and so on.
The question is, how will this impact performance of fetching ?
And do we care, as the original reason for PR #188 was about being able to fetch the specific SHA, not about optimizing future fetches.

So if we keep performance out-of the loop, as this was never the intention, we can basically wipe whole `refs/west` as part of `west update`.
Or consider to just keep the refs currently pointed to be `manifest`, so that if manifest points to `my_branch` we keep that refs, as it can be useful for, eg. :
https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1437
https://github.com/zephyrproject-rtos/west/issues/338 (have some rough ideas on using `refs/west` for checkpoints during a `west rebase`, that are wiped when rebase has completed).
Does that sound like an improvement ?

> Beyond not being able to collect the refs when they are no longer needed, it just feels weird.

Guess I can never take that feeling away from you :(

> and at least personally I'm bothered by all the `west/xyz` output I'm seeing showing up in `git log --oneline` output, for example. (E.g. all the backport branches if you have a downstream where zephyr isn't the manifest repository.)

Agree, valid point.
And cleaning out the parts of `refs/west` we don't actually need will for sure improve this.
Only reason I didn't do that in PR #188, was because I considered, `Why wipe it, if it improves performance of later fetches`.
So i'm fine with cleaning up the `refs/west` space, and then let `git gc` decide for how long potential SHAs are available.
It could mean additional fetches when switching between two very different manifest file, as example, switch between NCS and Zephyr as manifest project.
But I assume any degrade in fetch performance in this case is acceptable, as the performance will be similar to not having `refs/west` at all.

> 
> Within the Nordic downstream, I have always created a tag at the old point in history before rewriting history in any repository managed by west. This ensures an explicit pointer to the ref is available in a place where a user can expect to find it. 

And that is still true. We don't touch that tag, and it will be available until user removes it.
The tag is available where the user expects to find it.

> This feels like going behind the user's back in a weird way.

I disagree, we are not modifying anything the user has done.
But agree, current behavior would keep an additional ref to the same revision and that is what shows up in `git log`, and that seems to bother you, and I do acknowledge that, so will look into improving this.
And doing a little cleanup, as proposed above should help  :-)

Reply:
@mbolivar Just got to think.
When we do a `revision: pull/<no>/head` I actually got to think that this revision is NOT in the normal git refs-space.
Github does something similar to our `refs/west/*` refs space, and you can see that, as you can use the full refs as `git fetch <url> refs/pull/<no>/head`.
So special ref spaces, such as `refs/west/*` is not so special. (but must be cleaned up)